docs/reference-manual/native-image/ReplayBundles.md

```diff
@@ -0,0 +1,179 @@
+                          (uses files from input directory)
+```
+
+As we can see, there are several components in a replay bundle that we need to describe in more detail.
+
+### `META-INF`:
+
+Since the bundle is also a regular jar-file we have a `META-INF` subdirectory with the familiar `MANIFEST.MF`. The
+bundle can be used like a regular jar-launcher (by running command `java -jar <imagename>.nirb.jar`) so that the
+application we build an image from is instead executed on the JVM. For that purpose the `MANIFEST.MF` specifies the
+`rbundle/Launcher` as main class.
+
+Here we also find `RBUNDLE.MF`. This file is specific to replay bundles. Its existence makes clear that this is no
+ordinary jar-file but a native image replay bundle. The file contains version information of the native image replay
+bundle format itself and also which GraalVM version was used to create the bundle. This can later be used to report a
+warning message if a bundle gets replayed with a GraalVM version different from the one used to create the bundle.
+
+### `input`:
+
+This directory contains the entire amount of information needed to redo the previous image build. The original
+class-path and module-path entries are placed into corresponding files (for jar-files) and subdirectories (for
+directory-based class/module-path entries) into the `input/classes/cp` (original -cp/--class-path entries) and the
+`input/classes/p` (original -p/--module-path entries) folders. The `input/stage` folder contains all information
+needed to replicate the previous build context.
+
+#### `input/stage`:
+
+Here we have `build.cmd` that contains all native-image command line options used in the previous build. Note that
+**even the initial build that created the bundle already uses a class- and/or module-path that refers to the contents
+of the `input/classes` folder**. This way we can guarantee that a replay build sees exactly the same relocated
+class/module-path entries as the initial build. The use of `run.cmd` is explained later.
+
+File `all.env` contains the environment variables that we allowed the builder to see during the initial build and
+`all.properties` the respective system-properties.
+
+### `build`:
+
+This folder is used to document the build process that lead to the image that was created alongside the bundle.
+The `report` sub-folder holds `build.log`. It is equivalent to what would have been created if the user had appended
+`|& tee build.log` to the original native-image command line. Additionally, we have several json-files:
+* `analysis_results.json`: Contains the results of the static analysis. A rerun should compare the new
+`analysis_results.json` file with this one and report deviations in a user-friendly way.
+* `build_artifacts.json`: Contains a list of the artifacts that got created during the initial build. As before,
+changes should be reported to the user. 
+* `build_output.json`: Similar information as `build.log`.
```

Comment:
I think `build_output.json` (is that `-H:BuildOutputJSONFile`?), doesn't include linked native libs from the JDK. Maybe that's also something worth collecting as it might be a source of differences.

Reply:
Are you suggesting to record which static native libs have been linked into the image by the native-linker?

I would consider that redundant as it is fully determined by the version of GraalVM + the native-image command line options.

Reply:
> Are you suggesting to record which static native libs have been linked into the image by the native-linker?

Yes, the ones coming from the JDK. For example:

```
$ strings ./debug-symbols-smoke-mandrel | grep com.oracle.svm.core.VM.Target.StaticLibraries
com.oracle.svm.core.VM.Target.StaticLibraries=liblibchelper.a|libnet.a|libnio.a|libjava.a|libfdlibm.a|libzip.a|libjvm.a
```

> I would consider that redundant as it is fully determined by the version of GraalVM + the native-image command line options.

I'm not sure, we've seen issues with - a supposedly headless AWT - image build, which did include a different set depending on the build host (`GraphicsEnvironment.isHeadless()`). See https://github.com/oracle/graal/issues/5119